src/ophyd_async/core/device_save_loader.py

```diff
@@ -0,0 +1,175 @@
+    named_values = {
+        key: value["setpoint"] for key, value in zip(selected_signals, selected_values)
+    }
+    # Ignored values place in with value None so we know which ones were ignored
+    named_values.update({key: None for key in ignore})
+    return named_values
+
+
+def walk_rw_signals(
+    device: Device, path_prefix: Optional[str] = ""
+) -> Dict[str, SignalRW[Any]]:
+    """
+    Get all the SignalRWs from a device and store them with their dotted attribute
+    paths in a dictionary. Used as part of saving and loading a device
+    Parameters
+    ----------
+    device : Device
+        Ophyd device to retrieve read-write signals from.
+
+    path_prefix : str
+        For internal use, leave blank when calling the method.
+
+    Returns
+    -------
+    SignalRWs : dict
+        A dictionary matching the string attribute path of a SignalRW with the
+        signal itself.
+
+        See Also
+    --------
+    :func:`ophyd_async.core.get_signal_values`
+    :func:`ophyd_async.core.save_to_yaml`
+
+    """
+
+    if not path_prefix:
+        path_prefix = ""
+
+    signals: Dict[str, SignalRW[Any]] = {}
+    for attr_name, attr in device.children():
+        dot_path = f"{path_prefix}{attr_name}"
+        if type(attr) is SignalRW:
+            signals[dot_path] = attr
+        attr_signals = walk_rw_signals(attr, path_prefix=dot_path + ".")
+        signals.update(attr_signals)
+    return signals
+
+
+def save_to_yaml(phases: Sequence[Dict[str, Any]], save_path: str) -> None:
+    """
+    Serialises and saves a phase or a set of phases of a device's SignalRW's to a
+    yaml file.
+
+    Parameters
+    ----------
+    phases : dict or list of dicts
+        The values to save. Each item in the list is a seperate phase used when loading
+        a device. In general this variable be the return value of `get_signal_values`.
+
+    save_path : str
+        Path of the yaml file to write to
+
+    See Also
+    --------
+    :func:`ophyd_async.core.walk_rw_signals`
+    :func:`ophyd_async.core.get_signal_values`
+    :func:`ophyd_async.core.load_from_yaml`
+    """
+    yaml.add_representer(np.ndarray, ndarray_representer, Dumper=yaml.Dumper)
+    with open(save_path, "w") as file:
+        yaml.dump(phases, file, Dumper=OphydDumper, default_flow_style=None)
+
+
+def load_from_yaml(
+    device: Device, save_path: str
+) -> Generator[Msg, None, Sequence[Dict[str, Any]]]:
+    """
+    Loads a phase or a set of phases from a yaml file and puts value to an Ophyd device
+
+    Parameters
+    ----------
+    device : Device
+        Ophyd device with read-write Signals to write values to from yaml file.
+
+    save_path : str
+        Path of the yaml file to load from
+
+    See Also
+    --------
+    :func:`ophyd_async.core.save_to_yaml`
+    """
+    with open(save_path, "r") as file:
+        data_by_phase: Sequence[Dict[str, Any]] = yaml.full_load(file)
```

Comment:
Can we split this function here please. `load_from_yaml(save_path: str) -> Sequence[Dict[str, Any]]:`, then a different plan `set_signal_values(signals:  Dict[str, SignalRW], values: Sequence[Dict[str, Any]]) -> Generator[Msg, None, None]:`  that takes `signals` (as provided by `walk_rw_signals`), and `values` (as provided by `load_from_yaml`) and does the `abs_set` logic you do below

Reply:
Okay, have done this rebased onto the last passing commit for add-detector-logic